core/modules/replica/Controller.cc

```diff
@@ -0,0 +1,837 @@
+    return _threads.size();
+}
+
+void Controller::stop() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "stop");
+
+    if (not isRunning()) { return; }
+
+    // IMPORTANT:
+    //
+    //   Never attempt running these operations within LOCK_GUARD
+    //   due to a possibile deadlock when asynchronous handlers will be
+    //   calling the thread-safe methods. A problem is that until they finish
+    //   in a clean way (as per the _work.reset()) the thread will never finish,
+    //   and the application will hang on _thread->join().
+
+    // LOCK_GUARD  (disabled)
+
+    // These steps will cancel all oustanding requests
+    _messenger->stop();
+
+    // Destoring this object will let the I/O service to (eventually) finish
+    // all on-going work and shut down all Controller's threads. In that case there
+    // is no need to stop the service explicitly (which is not a good idea anyway
+    // because there may be outstanding synchronous requests, in which case the service
+    // would get into an unpredictanle state.)
+
+    _work.reset();
+
+    // At this point all outstanding requests should finish and all threads
+    // should stop as well.
+    join();
+
+    // Always do so in order to put service into a clean state. This will prepare
+    // it for further usage.
+    _io_service.reset();
+
+    _threads.clear();
+
+    // Double check that the collection of requests is empty.
+
+    if (not _registry.empty()) {
+        throw std::logic_error(
+            "Controller::stop() the collection of outstanding requests is not empty");
+    }
+}
+
+void Controller::join() {
+    for (auto&& t: _threads) { t->join(); }
+}
+
+ReplicationRequest::pointer Controller::replicate(
+                                std::string const& workerName,
+                                std::string const& sourceWorkerName,
+                                std::string const& database,
+                                unsigned int chunk,
+                                ReplicationRequest::callback_type onFinish,
+                                int  priority,
+                                bool keepTracking,
+                                bool allowDuplicate,
+                                std::string const& jobId,
+                                unsigned int requestExpirationIvalSec) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "replicate");
+
+    LOCK_GUARD;
+
+    assertIsRunning();
+
+    Controller::pointer controller = shared_from_this();
+
+    ReplicationRequest::pointer request =
+        ReplicationRequest::create(
+            _serviceProvider,
+            _io_service,
+            workerName,
+            sourceWorkerName,
+            database,
+            chunk,
+            [controller] (ReplicationRequest::pointer request) {
+                controller->finish(request->id());
+            },
+            priority,
+            keepTracking,
+            allowDuplicate
+            ,_messenger
```

Comment:
there's a few places where you put the comma before the last argument (3 in this file that are `,messenger`). I find it odd to read, but are you doing this intentionally?

Reply:
Thank you for noticing this! It's an artifact of the prior option of having two bases of requests, the older one didn't have the current Messenger. Hence, a switch was made via a macro. Something like this:

...
(p1,
p2
#ifdef MESSENGER_BASED
,messenger
#endif
)

That's fixed now.